addons/web_editor/static/src/js/wysiwyg/MoveNodePlugin.js

```diff
@@ -0,0 +1,374 @@
+    }
+    _updateAnchorWidgets(newAnchorWidget) {
+        let movableElement = newAnchorWidget && closestElement(newAnchorWidget, {
+            selector: allowedElements,
+            predicate: isNodeMovable,
+        });
+        // Retrive the first list container from the ancestors.
+        const listContainer = movableElement && ancestors(movableElement, this._editable)
+            .reverse()
+            .find(n => ['UL', 'OL'].includes(n.tagName));
+        movableElement = listContainer || movableElement;
+        if (movableElement && (movableElement !== this._currentMovableElement)) {
+            this._setMovableElement(movableElement);
+        }
+    }
+    _getMovableElements() {
+        return [...new Set([...this._editable.querySelectorAll(allowedElements)])]
+            .filter((node) =>
+                node.parentElement && node.parentElement.isContentEditable &&
+                isNodeMovable(node)
+            );
+    }
+    _getDroppableElements(draggableNode) {
+        return this._getMovableElements().filter((node) =>
+            !closestElement(node.parentElement, { predicate: (n) => n === draggableNode })
+        );
+    }
+    _setMovableElement(movableElement) {
+        this._removeMoveWidget && this._removeMoveWidget();
+        this._currentMovableElement = movableElement;
+        this._editor.disableAvatarForElement(movableElement);
+
+        const containerRect = this._widgetContainer.getBoundingClientRect();
+        const anchorBlockRect = this._currentMovableElement.getBoundingClientRect();
+        const closestList = closestElement(this._currentMovableElement, { selector: 'ul, ol' }); // Prevent overlap bullets.
+        const anchorX = closestList ? closestList.getBoundingClientRect().x : anchorBlockRect.x;
+        let anchorY = anchorBlockRect.y;
+        if (this._currentMovableElement.tagName.match(/H[1-6]/)) {
+            anchorY += (anchorBlockRect.height - widgetMoveSize) / 2;
+        }
+
+        this._moveWidget = this._document.createElement('div');
+        this._moveWidget.className = 'oe-sidewidget-move fa fa-sort';
+        this._widgetContainer.append(this._moveWidget);
+
+        let moveWidgetOffsetTop = 0;
+        if (movableElement.tagName === 'HR') {
+            const style = getComputedStyle(movableElement);
+            moveWidgetOffsetTop = parseInt(style.marginTop, 10) || 0;
+        }
+
+        this._moveWidget.style.width = `${widgetMoveSize}px`;
+        this._moveWidget.style.height = `${widgetMoveSize}px`;
+        this._moveWidget.style.top = `${anchorY - containerRect.y - moveWidgetOffsetTop}px`;
+        this._moveWidget.style.left = `${anchorX - containerRect.x - widgetContainerWidth}px`;
+        const timeout = setTimeout(()=> {
+            return this._moveWidget.classList.add('oe-sidewidget-move--show');
+        });
+
+        // Keep a reference for the closure.
+        const moveWidget = this._moveWidget;
+        this._removeMoveWidget = () => {
```

Comment:
I don't understand why we need to programmatically write a new method on `this`. If I understand correctly, it's to be sure to call it at some points ? Like basically everywhere we call it it's like "oh, right, if we had this thing going on, we need to cancel it", right ? I guess it's somewhat ok then... The first alternative I see would be to have a lot of variables on `this` to store the information that you already have here thanks to scoping so it wouldn't be better. Another alternative I see is to have a reference on `this` that stores all those callbacks, something like `_cancellableMoveWidgets` and you'd have a (haha again) one-time method `cancelMoveWidgets` which would cycle on those and call them all, but I guess you can only have one at a time anyway so it would be overkill. So, for those reasons, I think this is probably a viable way to do it.